File: android/app/src/main/java/com/danor93/RNMLKitPoseDetectionDemo/PoseDetector/PoseDetectorFrameProcessorPlugin.kt
/**
 * PoseDetectorFrameProcessorPlugin
 *
 * Author: Danor S.O.D.A
 * Last Edited: 26-11-2024
 *
 * This file implements a custom frame processor plugin for detecting human poses using Google's ML Kit.
 * The plugin integrates with VisionCamera's frame processing system to analyze frames for pose landmarks.
 * It maps the detected pose landmarks to their respective x and y coordinates in the image space, 
 * returning them in a structured format for React Native.
 */

 package com.danor93.RNMLKitPoseDetectionDemo.PoseDetector

 import android.media.Image
 import com.facebook.react.bridge.WritableNativeMap
 import com.google.android.gms.tasks.Tasks
 import com.google.mlkit.vision.common.InputImage
 import com.google.mlkit.vision.pose.Pose
 import com.google.mlkit.vision.pose.PoseDetection
 import com.google.mlkit.vision.pose.PoseLandmark
 import com.google.mlkit.vision.pose.defaults.PoseDetectorOptions
 import com.mrousavy.camera.frameprocessors.Frame
 import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
 import com.mrousavy.camera.frameprocessors.VisionCameraProxy
 import com.mrousavy.camera.core.types.Orientation
 import java.util.HashMap
 
 /**
  * PoseDetectorFrameProcessorPlugin
  *
  * A plugin for processing camera frames to detect human pose landmarks.
  *
  * @param proxy The VisionCamera proxy instance for frame processing.
  * @param options Additional options for configuring pose detection.
  */
 class PoseDetectorFrameProcessorPlugin(proxy: VisionCameraProxy, options: Map<String, Any>?) : FrameProcessorPlugin() {
 
     // Create a single instance of PoseDetector to reuse
     private val poseDetector = PoseDetection.getClient(
         PoseDetectorOptions.Builder()
             .setDetectorMode(PoseDetectorOptions.STREAM_MODE) // Always use STREAM_MODE for real-time
             .setPreferredHardwareConfigs(PoseDetectorOptions.CPU_GPU)
             .build()
     )
 
     // Track if detector is busy
     private var isProcessing = false
 
     /**
      * Processes the given camera frame to detect pose landmarks.
      *
      * @param frame The camera frame containing the image data.
      * @param arguments Additional arguments for configuring detection modes.
      * @return A HashMap containing detected pose landmarks with their coordinates.
      */
     override fun callback(frame: Frame, arguments: Map<String, Any>?): HashMap<String, Any> {
         // Skip frame if detector is busy
         if (isProcessing) {
             return HashMap()
         }
 
         try {
             isProcessing = true
 
             val mediaImage: Image = frame.image ?: throw Exception("No image available in the frame.")
             val orientation: Orientation = Orientation.fromUnionValue(frame.orientation.toString())
             
             // Create InputImage with YUV_420_888 format for better performance
             val image = InputImage.fromMediaImage(mediaImage, orientation.toSurfaceRotation())
             
             // Process pose detection
             val pose: Pose = Tasks.await(poseDetector.process(image))
 
             // Prepare the map to store pose landmarks
             val map = WritableNativeMap()
 
             /**
              * Adds a detected pose landmark to the result map.
              *
              * @param landmark The detected PoseLandmark.
              * @param landmarkName The name of the landmark as a key in the map.
              */
             fun addLandmarkToMap(landmark: PoseLandmark?, landmarkName: String) {
                 val landmarkMap = WritableNativeMap()
                 landmarkMap.putDouble("x", landmark?.position?.x?.toDouble() ?: 0.0)
                 landmarkMap.putDouble("y", landmark?.position?.y?.toDouble() ?: 0.0)
                 map.putMap(landmarkName, landmarkMap)
             }
 
             // Add all pose landmarks to the result map
             if (pose.allPoseLandmarks.isNotEmpty()) {
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_SHOULDER), "leftShoulder")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_SHOULDER), "rightShoulder")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_ELBOW), "leftElbow")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_ELBOW), "rightElbow")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_WRIST), "leftWrist")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_WRIST), "rightWrist")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_HIP), "leftHip")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_HIP), "rightHip")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_KNEE), "leftKnee")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_KNEE), "rightKnee")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_ANKLE), "leftAnkle")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_ANKLE), "rightAnkle")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_PINKY), "leftPinky")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_PINKY), "rightPinky")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_INDEX), "leftIndex")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_INDEX), "rightIndex")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_THUMB), "leftThumb")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_THUMB), "rightThumb")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_HEEL), "leftHeel")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_HEEL), "rightHeel")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_FOOT_INDEX), "leftFootIndex")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_FOOT_INDEX), "rightFootIndex")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.NOSE), "nose")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_EYE_INNER), "leftEyeInner")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_EYE), "leftEye")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_EYE_OUTER), "leftEyeOuter")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_EYE_INNER), "rightEyeInner")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_EYE), "rightEye")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_EYE_OUTER), "rightEyeOuter")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_EAR), "leftEar")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_EAR), "rightEar")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.LEFT_MOUTH), "leftMouth")
                 addLandmarkToMap(pose.getPoseLandmark(PoseLandmark.RIGHT_MOUTH), "rightMouth")
             }
 
             return map.toHashMap()
         } catch (e: Exception) {
             throw Exception("Error processing pose detection: ${e.message}")
         } finally {
             isProcessing = false
         }
     }
 }
 

File: ios/PoseDetector/PoseDetectorFrameProcessorPlugin.swift
//
//  PoseDetectorFrameProcessorPlugin.swift
//  RNMLKitPoseDetectionDemo
//
//  Author: Danor S.O.D.A
//  Last Edited: 26/11/2024
//
//  This file defines a custom frame processor plugin for detecting human poses using ML Kit.
//  It integrates with VisionCamera to process frames and extract pose landmarks in real time.
//  The detected landmarks are returned as x and y coordinates within a dictionary structure.
//

import MLKit
import VisionCamera

/// PoseDetectorFrameProcessorPlugin
///
/// A VisionCamera plugin leveraging ML Kit's Pose Detector to detect human poses in real-time frames.
/// It extracts pose landmarks with a confidence threshold and returns them as key-value pairs.
@objc(PoseDetectorFrameProcessorPlugin)
public class PoseDetectorFrameProcessorPlugin: FrameProcessorPlugin {
    // Pose Detector Options: Configured for optimized streaming mode
    private lazy var options: PoseDetectorOptions = {
        let options = PoseDetectorOptions()
        options.detectorMode = .stream
        return options
    }()
    
    // Lazy initialization of the Pose Detector
    private lazy var poseDetector: PoseDetector = {
        return PoseDetector.poseDetector(options: options)
    }()

    /// Initializes the PoseDetectorFrameProcessorPlugin with the VisionCamera proxy holder and optional configurations.
    ///
    /// - Parameters:
    ///   - proxy: The VisionCamera proxy holder.
    ///   - options: Optional configurations for custom detection settings.
    public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable : Any]! = [:]) {
        super.init(proxy: proxy, options: options)
    }

    /// Processes the camera frame to detect human poses and landmarks.
    ///
    /// - Parameters:
    ///   - frame: The frame containing image data and its orientation.
    ///   - arguments: Additional parameters to adjust processing behavior.
    /// - Returns: A dictionary containing pose landmarks with their x and y coordinates.
    public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable : Any]?) -> Any {
        let visionImage = VisionImage(buffer: frame.buffer)
        visionImage.orientation = frame.orientation

        var result: [String: Any] = [:]
        do {
            // Synchronously detect poses in the current frame
            let poses = try poseDetector.results(in: visionImage)
            
            // Process only the first detected pose for better performance
            if let pose = poses.first {
                PoseLandmarkWrapper.allCases.forEach { landmarkType in
                    if let landmarkData = getLandmarkData(pose: pose, type: landmarkType) {
                        result[landmarkType.description] = landmarkData
                    }
                }
            }
        } catch {
            print("Pose detection error: \(error.localizedDescription)")
        }

        return result
    }

    /// Extracts coordinates for a specific pose landmark if its confidence is sufficient.
    ///
    /// - Parameters:
    ///   - pose: The detected pose containing all landmarks.
    ///   - type: The specific landmark type to retrieve data for.
    /// - Returns: A dictionary with `x` and `y` coordinates if likelihood is above 50%; otherwise, nil.
    private func getLandmarkData(pose: Pose, type: PoseLandmarkWrapper) -> [String: CGFloat]? {
        let landmark = pose.landmark(ofType: type.poseLandmarkType)
        guard landmark.inFrameLikelihood > 0.5 else { return nil }
        return ["x": landmark.position.x, "y": landmark.position.y]
    }
}

/// PoseLandmarkWrapper: Enum encapsulating all possible PoseLandmarkTypes.
///
/// Provides utility to iterate over all ML Kit Pose landmarks, map them to PoseLandmarkType,
/// and generate string descriptions for debugging or dictionary keys.
private enum PoseLandmarkWrapper: CaseIterable {
    case nose, leftEyeInner, leftEye, leftEyeOuter,
         rightEyeInner, rightEye, rightEyeOuter,
         leftEar, rightEar,
         leftShoulder, rightShoulder,
         leftElbow, rightElbow,
         leftWrist, rightWrist,
         leftPinkyFinger, rightPinkyFinger,
         leftIndexFinger, rightIndexFinger,
         leftThumb, rightThumb,
         leftHip, rightHip,
         leftKnee, rightKnee,
         leftAnkle, rightAnkle,
         leftHeel, rightHeel,
         leftToe, rightToe

    /// Maps the enum case to its corresponding PoseLandmarkType.
    var poseLandmarkType: PoseLandmarkType {
        switch self {
        case .nose: return .nose
        case .leftEyeInner: return .leftEyeInner
        case .leftEye: return .leftEye
        case .leftEyeOuter: return .leftEyeOuter
        case .rightEyeInner: return .rightEyeInner
        case .rightEye: return .rightEye
        case .rightEyeOuter: return .rightEyeOuter
        case .leftEar: return .leftEar
        case .rightEar: return .rightEar
        case .leftShoulder: return .leftShoulder
        case .rightShoulder: return .rightShoulder
        case .leftElbow: return .leftElbow
        case .rightElbow: return .rightElbow
        case .leftWrist: return .leftWrist
        case .rightWrist: return .rightWrist
        case .leftPinkyFinger: return .leftPinkyFinger
        case .rightPinkyFinger: return .rightPinkyFinger
        case .leftIndexFinger: return .leftIndexFinger
        case .rightIndexFinger: return .rightIndexFinger
        case .leftThumb: return .leftThumb
        case .rightThumb: return .rightThumb
        case .leftHip: return .leftHip
        case .rightHip: return .rightHip
        case .leftKnee: return .leftKnee
        case .rightKnee: return .rightKnee
        case .leftAnkle: return .leftAnkle
        case .rightAnkle: return .rightAnkle
        case .leftHeel: return .leftHeel
        case .rightHeel: return .rightHeel
        case .leftToe: return .leftToe
        case .rightToe: return .rightToe
        }
    }

    /// Generates a human-readable description of the landmark type.
    var description: String {
        return String(describing: self)
    }
}
